source/hash_table.c

```diff
@@ -0,0 +1,582 @@
+    for (uint64_t i = 0; i <= max_disp; i++) {
+        printf("Displacement %2lld: %zu entries\n", i, disp_counts[i]);
+    }
+    free(disp_counts);
+    printf("\n");
+}
+#endif
+
+// Given a header template, allocates space for a hash table of the appropriate size, and copies the state header
+// into this allocated memory, which is returned.
+static struct hash_table_state *alloc_state(const struct hash_table_state *template) {
+    size_t elemsize;
+
+    if (!aws_common_mul_size_checked(template->size - 1, sizeof(template->slots[0]), &elemsize)) {
+        return NULL;
+    }
+
+    size_t size = elemsize + sizeof(*template);
+
+    if (size < elemsize) {
+        return NULL;
+    }
+
+    struct hash_table_state *state = aws_mem_acquire(template->alloc, size);
+
+    if (state == NULL) {
+        return state;
+    }
+
+    memcpy(state, template, sizeof(*template));
+    memset(&state->slots[0], 0, size - sizeof(*state) + sizeof(state->slots[0]));
+
+    return state;
+}
+
+// Computes the correct size and max_load based on a requested size.
+static int update_template_size(struct hash_table_state *template, size_t expected_elements) {
+    size_t min_size = expected_elements;
+
+    if (min_size < 2) {
+        min_size = 2;
+    }
+
+    size_t mask = ~(size_t)0, size = 1;
+    while (size < min_size) {
+        size = size << 1;
+        mask = mask << 1;
+
+        if (size == 0) {
+            // Overflow
+            return AWS_ERROR_OOM;
+        }
+    }
+    mask = ~mask;
+
+    // Cross-check - make sure we didn't just overflow somehow.
+    if (size < expected_elements) {
+        return AWS_ERROR_OOM;
+    }
+
+    template->size = size;
+    template->max_load = template->max_load_factor * template->size;
+    if (template->max_load >= size) {
+        template->max_load = size - 1;
+    }
+
+    // Make sure we don't overflow when computing memory requirements either
+    size_t required_mem = aws_common_mul_size_saturating(template->size, sizeof(struct hash_table_entry));
+    if (required_mem == SIZE_MAX || (required_mem + sizeof(struct hash_table_state)) < required_mem) {
+        return AWS_ERROR_OOM;
+    }
+
+    template->size = size;
+    template->mask = mask;
+
+    return AWS_ERROR_SUCCESS;
+}
+
+int aws_common_hash_table_init(struct aws_common_hash_table *map,
+    struct aws_allocator *alloc, size_t size,
+    aws_hash_fn_t hash_fn, aws_equals_fn_t equals_fn, aws_hash_element_destroy_t destroy_fn) {
+
+    struct hash_table_state template;
+    template.hash_fn = hash_fn;
+    template.equals_fn = equals_fn;
+    template.destroy_fn = destroy_fn;
+    template.alloc = alloc;
+
+    template.entry_count = 0;
+    template.max_load_factor = 0.95; // TODO - make configurable?
+
+    update_template_size(&template, size);
+    map->pImpl = alloc_state(&template);
+
+    if (!map->pImpl) {
+        return AWS_ERROR_OOM;
```

Comment:
These should be using aws_raise_error

Reply:
also alloc_state can return null if the requested size is too large. do we need a more granular error code for that?

Reply:
I don't think a more granular error is needed, the semantics of `AWS_ERROR_OOM` or `NULL` return from allocate are already "You asked for more memory than I can give you"

Reply:
Fair enough